avax-synapse-stablepool/contracts/Strategy.sol

```diff
@@ -0,0 +1,479 @@
+            _amount = _amountOfCrv.mul(virtualPrice).div(1e18);
+        }
+
+        if (amountFromVault > 0) {
+            yvToken.withdraw(amountFromVault);
+            if (withdrawProtection) {
+                //this tests that we liquidated all of the expected ytokens. Without it if we get back less then will mark it is loss
+                require(yBalance.sub(yvToken.balanceOf(address(this))) >= amountFromVault.sub(1), "YVAULTWITHDRAWFAILED");
+            }
+        }
+
+        uint256 toWithdraw = emissionToken.balanceOf(address(this)).sub(lpTokenBeforeBalance);
+
+        if (toWithdraw > 0) {
+            //if we have less than 18 decimals we need to lower the amount out
+            uint256 maxSlippage = toWithdraw.mul(DENOMINATOR.sub(slippageProtectionOut)).div(DENOMINATOR);
+            if(want_decimals < 18){
+                maxSlippage = maxSlippage.div(10 ** (uint256(uint8(18) - want_decimals)));
+            }
+
+            swapPool.removeLiquidityOneToken(toWithdraw, curveId, 0, now);
+        }
+
+        uint256 diff = IERC20(targetToken).balanceOf(address(this)).sub(wantBalanceBefore);
+
+        if (diff > _amount) {
+            _liquidatedAmount = _amount;
+        } else {
+            _liquidatedAmount = diff;
+            _loss = _amount.sub(diff);
+        }
+    }
+
+    function liquidatePosition(uint256 _amountNeeded)
+        internal
+        override
+        returns (uint256 _liquidatedAmount, uint256 _loss)
+    {
+        uint256 wantBal = IERC20(targetToken).balanceOf(address(this));
+
+        if (wantBal < _amountNeeded) {
+            (_liquidatedAmount, _loss) = withdrawSome(_amountNeeded.sub(wantBal));
+        }
+
+        _liquidatedAmount = Math.min(_amountNeeded, _liquidatedAmount.add(wantBal));
+    }
+
+    function liquidateAllPositions() internal override returns (uint256 _amountFreed) {
+        (_amountFreed, ) = liquidatePosition(1e36); //we can request a lot. dont use max because of overflow
+    }
+
+    function adjustPosition(uint256 _debtOutstanding) internal override {
+        if (lastInvest.add(minTimePerInvest) > block.timestamp) {
+            return;
+        }
+
+        // invest the rest of the want
+
+        uint256 _wantToInvest = Math.min(IERC20(targetToken).balanceOf(address(this)), maxSingleInvest);
+        if (_wantToInvest == 0) {
+            return;
+        }
+
+        uint256 expectedOut = _wantToInvest.mul(1e18).div(virtualPriceToWant());
+        uint256 maxSlip = expectedOut.mul(DENOMINATOR.sub(slippageProtectionIn)).div(DENOMINATOR);
+
+        // invest all the tokens we have
+        uint256[] memory data = new uint256[](4);
+        data[0] = nusd.balanceOf(address(this));
+        data[1] = dai.balanceOf(address(this));
+        data[2] = usdc.balanceOf(address(this));
+        data[3] = usdt.balanceOf(address(this));
+
+        if (data[0] > 0 || data[1] > 0 || data[2] > 0 || data[3] > 0) {
```

Comment:
This strategy can't assume what token will be deposited or in which pool we're gonna use it. It should be as generic as possible. Please refer to this [version](https://github.com/Tesseract-Finance/tesseract-strategies/blob/9357074271a7f37dfd20e2578d1d74224dcc7c95/synnapse/contracts/Strategy.sol#L385)